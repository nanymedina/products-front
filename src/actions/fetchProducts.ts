import axios from 'axios';
import { Dispatch } from '@src/type/action';

const fetchProducts = async (search: string, dispatch: Dispatch): Promise<any> => {
  try {
    const response = await axios.post(process.env.GRAPHQL_URL || '', {
      query: `
      query {
        getProducts(search:"${search}") {
          id
          brand
          description
          price
          oldPrice
          discountRate
          image
        }
      }`,
    });

    const { getProducts: products } = response.data.data;

    dispatch({ type: 'FETCH_PRODUCTS', params: { products, search } });
  } catch (e) {
    dispatch({ type: 'ERROR_FETCH_PRODUCTS', params: { error: e.message } });
  }
};

export default fetchProducts;
