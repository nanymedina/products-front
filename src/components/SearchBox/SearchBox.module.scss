.search_box {
  width: 500px;
  height: 45px;
  padding: 0 20px;
  margin: 0;
  border-radius: 25px;
  border: 1px solid #2d9fd9;
  background-color: #fff;
  font-weight: 400;
}

.search_box:focus {
  outline: none;
}
