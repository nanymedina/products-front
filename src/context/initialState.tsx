import StateT from '@src/type/state';

const initialState: StateT = {
  products: [],
  search: null,
  shoppingCart: [],
  error: null,
};

export default initialState;
