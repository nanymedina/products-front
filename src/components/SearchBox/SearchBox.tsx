import React, { ReactElement } from 'react';

import { useDispatchApp } from '@src/context/appContext';
import fetchProducts from '@src/actions/fetchProducts';
import resetProducts from '@src/actions/restProducts';

import Style from './SearchBox.module.scss';

type SearcherT = {
  placeholder?: string;
};

const SearchBox = ({ placeholder, ...props }: SearcherT): ReactElement => {
  const dispatch = useDispatchApp();

  const handleOnKeyPress = (event: any) => {
    if (event.charCode === 13) {
      const search = event.target.value;
      search.length > 0 && fetchProducts(search, dispatch);
    }
  };

  const handleOnChange = (event: any) => {
    const search = event.target.value;
    search.length === 0 && resetProducts(dispatch);
  };

  return (
    <input
      data-testid="search-box"
      className={Style.search_box}
      type="text"
      onKeyPress={handleOnKeyPress}
      placeholder={placeholder}
      onChange={handleOnChange}
      {...props}
    />
  );
};

export default SearchBox;
