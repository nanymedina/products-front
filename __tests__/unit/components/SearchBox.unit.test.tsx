import { render, fireEvent } from '@testing-library/react';
import SearchBox from '@components/SearchBox';
import { useDispatchApp } from '@src/context/appContext';
import fetchProducts from '@src/actions/fetchProducts';
import resetProducts from '@src/actions/restProducts';
import { act } from 'react-dom/test-utils';

jest.mock('@src/context/appContext');
jest.mock('@src/actions/fetchProducts');
jest.mock('@@src/actions/restProducts');

describe('Search Box', () => {
  const useDispatchAppMock = useDispatchApp as jest.Mock;
  const dispatch = jest.fn();
  useDispatchAppMock.mockReturnValue(dispatch);
  const fetchProductsMock = fetchProducts as jest.Mock;
  const resetProductsMock = resetProducts as jest.Mock;

  afterEach(() => {
    useDispatchAppMock.mockClear();
    fetchProductsMock.mockClear();
    resetProductsMock.mockClear();
  });

  it('dispatch fetch product with has search', () => {
    const { getByTestId } = render(<SearchBox />);
    const searchBox = getByTestId('search-box');

    fireEvent.change(searchBox, { target: { value: 'a' } });
    fireEvent.change(searchBox, { target: { value: 'as' } });
    fireEvent.keyPress(searchBox, { key: 'Enter', code: 13, charCode: 13 });

    expect(fetchProductsMock).toHaveBeenCalledTimes(1);
    expect(resetProductsMock).not.toBeCalled();
  });

  xit('dispatch reset search with has search', () => {
    const { getByTestId } = render(<SearchBox />);
    const searchBox = getByTestId('search-box');
    act(() => {
      fireEvent.change(searchBox, { target: { value: '' } });
      fireEvent.keyPress(searchBox, { key: 'Enter', code: 13, charCode: 13 });
    });
    expect(fetchProductsMock).not.toBeCalled();
    expect(resetProductsMock).toBeCalled();
  });

  it('should show search box with placeholder', () => {
    const { getByTestId } = render(<SearchBox placeholder="placeholder" />);
    const searchBox = getByTestId('search-box');

    expect(searchBox).toMatchSnapshot();
  });

  it('should show search box without placeholder', () => {
    const { getByTestId } = render(<SearchBox />);
    const searchBox = getByTestId('search-box');

    expect(searchBox).toMatchSnapshot();
  });
});
