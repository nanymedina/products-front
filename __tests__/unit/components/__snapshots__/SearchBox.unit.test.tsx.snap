// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Box should show search box with placeholder 1`] = `
<input
  data-testid="search-box"
  placeholder="placeholder"
  type="text"
/>
`;

exports[`Search Box should show search box without placeholder 1`] = `
<input
  data-testid="search-box"
  type="text"
/>
`;
